File: app/src/main/java/com/tunjid/fingergestures/fragments/AppFragment.java
package com.tunjid.fingergestures.fragments;


import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.theartofdev.edmodo.cropper.CropImage;
import com.tunjid.androidbootstrap.core.abstractclasses.BaseFragment;
import com.tunjid.androidbootstrap.recyclerview.ListManager;
import com.tunjid.androidbootstrap.recyclerview.ListManagerBuilder;
import com.tunjid.fingergestures.App;
import com.tunjid.fingergestures.BackgroundManager;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.adapters.AppAdapter;
import com.tunjid.fingergestures.baseclasses.MainActivityFragment;
import com.tunjid.fingergestures.viewholders.AppViewHolder;
import com.tunjid.fingergestures.viewmodels.AppViewModel;

import java.io.File;
import java.util.Arrays;
import java.util.stream.IntStream;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentTransaction;

import static com.tunjid.fingergestures.BackgroundManager.DAY_WALLPAPER_PICK_CODE;
import static com.tunjid.fingergestures.BackgroundManager.NIGHT_WALLPAPER_PICK_CODE;
import static java.lang.Math.abs;

public class AppFragment extends MainActivityFragment
        implements
        AppAdapter.AppAdapterListener {

    private static final String ARGS_ITEM = "items";
    private static final String IMAGE_SELECTION = "image/*";

    private int[] items;
    private ListManager<AppViewHolder, Void> listManager;

    public static AppFragment newInstance(int[] items) {
        AppFragment fragment = new AppFragment();
        Bundle args = new Bundle();

        args.putIntArray(ARGS_ITEM, items);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public String getStableTag() {
        return Arrays.toString(getArguments().getIntArray(ARGS_ITEM));
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        items = getArguments().getIntArray(ARGS_ITEM);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ViewGroup root = (ViewGroup) inflater.inflate(R.layout.fragment_home, container, false);

        listManager = new ListManagerBuilder<AppViewHolder, Void>()
                .withRecyclerView(root.findViewById(R.id.options_list))
                .withAdapter(new AppAdapter(items, this))
                .withLinearLayoutManager()
                .addDecoration(divider())
                .addScrollListener((dx, dy) -> { if (abs(dy) > 3) toggleToolbar(dy < 0); })
                .build();

        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onResume() {
        super.onResume();
        notifyDataSetChanged();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        toggleToolbar(true);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        Activity activity = getActivity();
        if (activity == null) return;

        if (resultCode != Activity.RESULT_OK) {
            showSnackbar(R.string.cancel_wallpaper);
        }
        else if (requestCode == DAY_WALLPAPER_PICK_CODE || requestCode == NIGHT_WALLPAPER_PICK_CODE) {
            cropImage(data.getData(), requestCode);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        listManager.clear();
        listManager = null;
    }

    @Override
    public void pickWallpaper(@BackgroundManager.WallpaperSelection int selection) {
        if (!App.hasStoragePermission()) showSnackbar(R.string.enable_storage_settings);
        else startActivityForResult(Intent.createChooser(new Intent()
                .setType(IMAGE_SELECTION)
                .setAction(Intent.ACTION_GET_CONTENT), ""), selection);
    }

    public int[] getItems() {
        return items;
    }

    public void notifyItemChanged(@AppViewModel.AdapterIndex int position) {
        int index = IntStream.range(0, items.length).filter(i -> items[i] == position).findFirst().orElse(-1);
        if (index != -1) listManager.notifyItemChanged(index);
    }

    public void notifyDataSetChanged() {
        listManager.notifyDataSetChanged();
    }

    @Nullable
    @Override
    @SuppressLint("CommitTransaction")
    @SuppressWarnings("ConstantConditions")
    public FragmentTransaction provideFragmentTransaction(BaseFragment fragmentTo) {
        return getFragmentManager().beginTransaction()
                .setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out,
                        android.R.anim.fade_in, android.R.anim.fade_out);
    }

    public void cropImage(Uri source, @BackgroundManager.WallpaperSelection int selection) {
        BackgroundManager backgroundManager = BackgroundManager.getInstance();
        int[] aspectRatio = backgroundManager.getScreenAspectRatio();
        if (aspectRatio == null) return;

        Activity activity = getActivity();
        if (activity == null) return;

        File file = backgroundManager.getWallpaperFile(selection, activity);
        Uri destination = Uri.fromFile(file);

        CropImage.activity(source)
                .setOutputUri(destination)
                .setFixAspectRatio(true)
                .setAspectRatio(aspectRatio[0], aspectRatio[1])
                .setMinCropWindowSize(100, 100)
                .setOutputCompressFormat(Bitmap.CompressFormat.PNG)
                .start(activity, this);
    }
}

File: app/src/main/java/com/tunjid/fingergestures/fragments/AppFragment.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.fragments


import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.tunjid.androidx.core.delegates.fragmentArgs
import com.tunjid.androidx.recyclerview.addScrollListener
import com.tunjid.androidx.recyclerview.verticalLayoutManager
import com.tunjid.fingergestures.models.uiState
import com.tunjid.fingergestures.models.updatePartial
import com.tunjid.fingergestures.R
import com.tunjid.fingergestures.adapters.appAdapter
import com.tunjid.fingergestures.baseclasses.divider
import com.tunjid.fingergestures.databinding.FragmentHomeBinding
import com.tunjid.fingergestures.di.activityViewModelFactory
import com.tunjid.fingergestures.mapDistinct
import com.tunjid.fingergestures.models.AppState
import com.tunjid.fingergestures.viewmodels.AppViewModel
import com.tunjid.fingergestures.viewmodels.Tab
import kotlin.math.abs

private var AppFragment.tab by fragmentArgs<Tab>()
private fun AppFragment.tabItems(state: AppState?) = state?.items?.filter { it.tab == tab }

class AppFragment : Fragment(R.layout.fragment_home) {

    private val viewModel by activityViewModelFactory<AppViewModel>()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val listAdapter = appAdapter(tabItems(viewModel.state.value))

        FragmentHomeBinding.bind(view).optionsList.apply {
            layoutManager = verticalLayoutManager()
            adapter = listAdapter
            itemAnimator = null

            addItemDecoration(divider())
            addScrollListener { _, dy -> if (abs(dy) > 3) ::uiState.updatePartial { copy(toolbarShows = dy < 0) } }
        }

        viewModel.state
            .mapDistinct(::tabItems)
            .observe(viewLifecycleOwner, listAdapter::submitList)
    }

    override fun onResume() {
        super.onResume()
        ::uiState.updatePartial { copy(toolbarShows = true) }
    }

    companion object {
        fun newInstance(tab: Tab): AppFragment = AppFragment().apply {
            this.tab = tab
        }
    }
}
